src/interfaces/plugins.ts

```diff
@@ -13,45 +13,114 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
 */
-
-const debug = require('debug')('signalk:interfaces:plugins')
-const fs = require('fs')
-const path = require('path')
-const express = require('express')
-const _ = require('lodash')
+import Debug from 'debug'
+import { Request, Response } from 'express'
+const debug = Debug('signalk:interfaces:plugins')
+// @ts-ignore
+import { getLogger } from '@signalk/streams/logging'
+import express from 'express'
+import fs from 'fs'
+import _ from 'lodash'
+import path from 'path'
+
+// tslint:disable-next-line:no-var-requires
 const modulesWithKeyword = require('../modules').modulesWithKeyword
-const getLogger = require('@signalk/streams/logging').getLogger
-const _putPath = require('../put').putPath
-const { getModulePublic } = require('../config/get')
-const { queryRequest } = require('../requestResponse')
+// tslint:disable-next-line:no-var-requires
+const put = require('../put')
+// tslint:disable-next-line
+const _putPath = put.putPath
+// tslint:disable-next-line:no-var-requires
+const getModulePublic = require('../config/get').getModulePublic
+// tslint:disable-next-line:no-var-requires
+const queryRequest = require('../requestResponse').queryRequest
 
 // #521 Returns path to load plugin-config assets.
 const getPluginConfigPublic = getModulePublic('@signalk/plugin-config')
 
-const DEFAULT_ENABLED_PLUGINS = process.env['DEFAULTENABLEDPLUGINS']
-  ? process.env['DEFAULTENABLEDPLUGINS'].split(',')
+const DEFAULT_ENABLED_PLUGINS = process.env.DEFAULTENABLEDPLUGINS
+  ? process.env.DEFAULTENABLEDPLUGINS.split(',')
   : []
 
-module.exports = function (app) {
-  const onStopHandlers = {}
+export type PluginFactory = (serverApi: ServerAPI) => Plugin
+
+export interface Plugin {
+  start: (config: any, restart: (newConfiguration: any) => void) => any
+  stop: () => any
+}
+
+interface PluginInfo extends Plugin {
+  enableLogging: any
+  packageName: any
+  packageLocation: string
+  registerWithRouter: any
+  signalKApiRoutes: any
+  name: string
+  id: string
+  schema: () => void | object
+  uiSchema: () => void | object
+  version: string
+  description: string
+  state: string
+  enabledByDefault: boolean
+  statusMessage: () => string | void
+}
+
+export interface ServerAPI {
+  getSelfPath: (path: string) => void
+  getPath: (path: string) => void
+  putSelfPath: (path: string) => void
+  // putPath,
+  // queryRequest,
+  error: (msg: string) => void
+  debug: (msg: string) => void
+  registerDeltaInputHandler: (handler: any) => void
+  setProviderStatus: (msg: string) => void
+  handleMessage: (id: string, msg: any) => void
```

Comment:
The naming is not the most fitting, but it is what it is: without all plugins changing we can only create aliases, renaming would break existing plugins that use `handleMessage`.

Reply:
Yes, but that is what I am suggesting (create an alias).  In the `doRegisterPlugin()` method, you have this:
`const appCopy: ServerAPI = _.assign({}, app, {
      getSelfPath,
      getPath,
      ...
`

Doesn't that mean that every method in `app` will be in appCopy? So, handleMessage() will always be there, even if it is not declared in the interface declaration.  I was suggesting that a new `outputDelta()` method be defined in the interface similar to the way you RE-define `setProviderStatus()` to take a single parameter vs. two.  `handleMessage()` would still be available to legacy plugins, but the new, "official" API would have the renamed method.

Reply:
Gotcha.

Actually I think with TypeScript we can start from scratch: the to-be-published * ts interfaces* need not have everything exactly in the way things are in JavaScript, we can choose to add some stuff and not include some. This way we can guide people to use the correct API and start deprecating old ones.

Reply:
Yes!  That is exactly what I am thinking.  Take this as an opportunity to look critically at the API, and based on past experience, decide what methods and techniques are the ones that are believed to be the "best practices" and define those in the interface.  Even things like `setProviderStatus()` might benefit from a "plugin centric" name like `setPluginStatus()` - at least as far as the interface is concerned.  If the `.streambundle` stuff should be considered "not desirable", don't bother putting it or any equivalent in the interface at all.